x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java

```diff
@@ -218,13 +218,28 @@ public T transformUp(Function<? super T, ? extends T> rule) {
             transformedChildren.add(next);
         }
 
-        return (childrenChanged ? replaceChildren(transformedChildren) : (T) this);
+        return (childrenChanged ? replaceEachChild(transformedChildren) : (T) this);
     }
 
     /**
-     * Replace the children of this node.
+     * Replaces each child of the current node and validates that the number of children before 
+     * and after this operation will stay the same (default invariant property of node transformations).
      */
-    public abstract T replaceChildren(List<T> newChildren);
+    public final T replaceEachChild(List<T> newChildren) {
+        // check for the invariant here once, instead of checking within every class
+        // invariant: before and after any transform the number of children should not change
+        if (newChildren.size() != children.size()) {
+            throw new QlIllegalArgumentException(
+                "Expected the same number of children [" + children.size() + "], but received [" + newChildren.size() + "]");
+        }
+        return replaceChildren(newChildren);
+    }
+
+    /**
+     * Constructs a new node, keeping the current properties of the node, but changing the list of 
+     * the children.
+     */
+    protected abstract T replaceChildren(List<T> newChildren);
```

Comment:
The method should be kept public just like `replaceChildren` above since it's used not just by subclasses but also by Analyzer and Optimizer.

Reply:
The places where this `replaceChildren` method is used:

1. `Node` class: protected is enough
2. test classes: protected is enough
3. `SimplifyConditional` and `SimplifyCase` rules in the `org.elasticsearch.xpack.sql.optimizer.Optimizer`: needs public access or wrapper, but only for the `Case` and `ArbitraryConditionalFunction` classes.

Other previous usages of `replaceChildren` in the Analyzer, AnalyzerRules, Optimizers, Pipes and LogicalPlanBuilder were all replaced by `pubilc replaceEachChild` calls. They don't need `public` access of the `Node.replaceChildren`.

Although there is no need for `public` visibility on the `Node` level, I can certainly turn `replaceChildren` method back to `public` to decrease the noise of this PR and to remove the promotions to `public` in case `#3` above.

Reply:
Myself, I don't object against making it protected, since the public variant is only required for a couple of subclasses.